train/NEAT/evolve.py

```diff
@@ -0,0 +1,157 @@
+    return info['distance']
+
+def eval_genome(genome, config):
+
+    net = neat.nn.FeedForwardNetwork.create(genome, config)
+    stuckMax = 600
+    
+    for i in range(0, 32):
+        observation = env[i].reset()
+        done = False
+        stuck = 0
+        while not done:
+            # Get move from NN
+            outputs = clean_outputs(net.activate(observation.flatten()))
+
+            # Make move
+            observation, reward, done, info = env[i].step(outputs)
+
+            # Check if Mario is progressing in level
+            stuck += 1 if reward <= 0 else 0
+
+            #TODO: Needs improvment, need to disable at end of level and when in a pipe.
+            if (stuck > stuckMax):
+                env[i].close()
+                return calculate_fitness(info)
+
+        if info['life'] == 0:
+            break
+
+        env[i].close()
+    return calculate_fitness(info)
+
+def run(config):
+    if args.checkpoint:
+        pop = neat.Checkpointer.restore_checkpoint(args.checkpoint)
+    else:
+        pop = neat.Population(config)
+
+    # Checkpoint every 5 generations or 10 min.
+    pop.add_reporter(neat.Checkpointer(5, 600))
+
+    pop.add_reporter(neat.StdOutReporter(True))
+    stats = neat.StatisticsReporter()
+    pop.add_reporter(stats)
+
+    pe = neat.ParallelEvaluator(args.numCores, eval_genome)
+    
+    # We could use this to set a limit of generations... But i dont think we need too... 
+    #for x in range(args.generations if args.generations > 0 else sys.maxsize): 
+    while True:
+        winner = pop.run(pe.evaluate, 5)
+        
+        if args.vineLogging:
+            #TODO: Fix Vine Logging, everything needed should be in stats. 
+            visualise.save_statistics(stats)
+        else:
+            visualise.plot_stats(stats, ylog=False, view=False)
+            visualise.plot_species(stats, view=False)
+
+        # Save the best Genome from the last 5 gens.
+        with open('Best-{}.pkl'.format(len(stats.most_fit_genomes)), 'wb') as output:
+            pickle.dump(winner, output, 1)
+
+def playBest(config, bestFile):
+    #TODO: Set to normal...
+    # Error - The mode "normal" is not supported. Supported options are "algo" or "human"
+    wrapper = SetPlayingMode('human')
+    e = wrapper(env[0])
+
+    genome = pickle.load(open(bestFile, 'rb'))
+
+    net = neat.nn.FeedForwardNetwork.create(genome, config)
+
+    done = False
+    observation = e.reset()
+    while not done:
+        action = clean_outputs(net.activate(observation.flatten()))
+        
+        print('Action taken by NN : [{} {} {} {} {} {}]'.format(
+                ('UP' if action[0] else ''),
+                ('Left' if action[1] else ''),
+                ('Right' if action[2] else ''),
+                ('Down' if action[3] else ''),
+                ('A' if action[4] else ''),
+                ('B' if action[5] else '')))
+                
+        e.step(action)
+
+def load_config(config_file):
```

Comment:
Could we not just pass in the absolute path to the config file and not do any relative path resolution?

Reply:
To be honest, i don't like have things so absolute. 
But whats the advantage of it? 
this works just fine when i am running within the docker container or directly on my machine.

Reply:
If we did ever run this from Wooey this would likely break - the advantage is that your config can be anywhere without needing them to be sitting directly next to your script.

Reply:
While i can see the advantage of using wooey in other situations, is it really bringing us anything?

Given that we really just run 1 script that can take days to complete, is the added complexity needed? 

To be honest, i never use it. I find it quicker to just login and run the script.